ddtrace/internal/runtime/lightstep_metrics.py

```diff
@@ -0,0 +1,215 @@
+import logging
+import os
+import platform
+import random
+import string
+
+from ... import _worker
+
+from .constants import GC_RUNTIME_METRICS
+from .metric_collectors import GCRuntimeMetricCollector, RuntimeMetricCollector
+from .runtime_metrics import RuntimeCollectorsIterable
+from google.protobuf.duration_pb2 import Duration
+from google.protobuf.timestamp_pb2 import Timestamp
+from ddtrace.vendor.lightstep.collector_pb2 import KeyValue, Reporter
+from ddtrace.vendor.lightstep.metrics_pb2 import IngestRequest, MetricKind
+
+log = logging.getLogger(__name__)
+
+
+LS_PROCESS_CPU_TIME_SYS = "process.cpu.system"
+LS_PROCESS_CPU_TIME_USER = "process.cpu.user"
+LS_PROCESS_MEM_RSS = "process.mem.rss"
+LS_SYSTEM_CPU_TIME_SYS = "cpu.system"
+LS_SYSTEM_CPU_TIME_USER = "cpu.user"
+LS_SYSTEM_CPU_TIME_IDLE = "cpu.idle"
+LS_SYSTEM_CPU_TIME_STEAL = "cpu.steal"
+LS_SYSTEM_CPU_TIME_NICE = "cpu.nice"
+LS_SYSTEM_MEM_AVAIL = "mem.available"
+LS_SYSTEM_MEM_USED = "mem.used"
+LS_SYSTEM_NET_RECV = "net.recv"
+LS_SYSTEM_NET_SENT = "net.sent"
+
+LS_RUNTIME_METRICS = set(
+    [
+        LS_PROCESS_CPU_TIME_SYS,
+        LS_PROCESS_CPU_TIME_USER,
+        LS_PROCESS_MEM_RSS,
+        LS_SYSTEM_CPU_TIME_SYS,
+        LS_SYSTEM_CPU_TIME_USER,
+        LS_SYSTEM_CPU_TIME_IDLE,
+        LS_SYSTEM_CPU_TIME_STEAL,
+        LS_SYSTEM_CPU_TIME_NICE,
+        LS_SYSTEM_MEM_AVAIL,
+        LS_SYSTEM_MEM_USED,
+        LS_SYSTEM_NET_RECV,
+        LS_SYSTEM_NET_SENT,
+    ]
+)
+
+
+class LightstepPSUtilRuntimeMetricCollector(RuntimeMetricCollector):
+    """Collector for psutil metrics.
+
+    Performs batched operations via proc.oneshot() to optimize the calls.
+    See https://psutil.readthedocs.io/en/latest/#psutil.Process.oneshot
+    for more information.
+    """
+
+    required_modules = ["ddtrace.vendor.psutil"]
+    stored_value = dict(
+        CPU_TIME_SYS_TOTAL=0,
+        CPU_TIME_USER_TOTAL=0,
+        SYSTEM_CPU_TOTAL=0,
+        SYSTEM_CPU_USER_TOTAL=0,
+        SYSTEM_CPU_IDLE_TOTAL=0,
+        SYSTEM_CPU_NICE_TOTAL=0,
+        NET_RECV_TOTAL=0,
+        NET_SENT_TOTAL=0,
+    )
+    previous_value = dict()
+
+    def _on_modules_load(self):
+        self.proc = self.modules["ddtrace.vendor.psutil"].Process(os.getpid())
+        self.cpu = self.modules["ddtrace.vendor.psutil"].cpu_times
+        self.mem = self.modules["ddtrace.vendor.psutil"].virtual_memory
+        self.net = self.modules["ddtrace.vendor.psutil"].net_io_counters
+
+    def collect_fn(self, keys):
+        with self.proc.oneshot():
+            # only return time deltas
+            # TODO[tahir]: better abstraction for metrics based on last value
```

Comment:
Remind Tahir to do this :slightly_smiling_face:

Reply:
Removed comment